File: leetcode/lru-cache/lru-cache.java
// https://leetcode.com/problems/lru-cache/

final class LRUCache {
    LRUCache(int capacity) {
        if (capacity <= 0)
            throw new IllegalArgumentException("capacity must be positive");

        _capacity = capacity;
    }

    int get(int key) {
        var node = _map.get(key);
        if (node == null) return NOT_FOUND;

        bump(node);
        return node.value();
    }

    void put(int key, int value) {
        var node = _map.get(key);

        if (node != null) {
            node.setValue(value);
            bump(node);
            return;
        }

        if (_map.size() < _capacity) {
            node = new Node(key, value, _sentinel);
        } else {
            node = _sentinel.prev();
            _map.remove(node.key());
            node.set(key, value);
            bump(node);
        }

        _map.put(key, node);
    }

    private static final class Node {
        Node() {
            _prev = _next = this;
            _key = _value = 0;
        }

        Node(int key, int value, Node prev) {
            attach(prev);
            _key = key;
            _value = value;
        }

        void connectAfter(Node prev) {
            detach();
            attach(prev);
        }

        Node prev() {
            return _prev;
        }

        Node next() {
            return _next;
        }

        int key() {
            return _key;
        }

        int value() {
            return _value;
        }

        void setValue(int value) {
            _value = value;
        }

        void set(int key, int value) {
            _key = key;
            _value = value;
        }

        private void detach() {
            _prev._next = _next;
            _next._prev = _prev;
        }

        private void attach(Node prev) {
            _prev = prev;
            _next = prev._next;
            _prev._next = _next._prev = this;
        }

        private Node _prev;

        private Node _next;

        private int _key;

        private int _value;
    }

    private static final int NOT_FOUND = -1;

    private void bump(Node node) {
        node.connectAfter(_sentinel);
    }

    private final Map<Integer, Node> _map = new HashMap<>(); // key -> node

    private final Node _sentinel = new Node();

    private final int _capacity;
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */

File: leetcode/lru-cache/lru-cache.py
# https://leetcode.com/problems/lru-cache


class _Node:
    """A helper doubly linked list node for LRUCache."""

    __slots__ = ('_prev', '_next', 'key', 'value')

    def __init__(self, key = None, value = None, prev : '_Node' = None):
        """Creates a new sentinel or non-sentinel node."""
        if prev is None:
            self._prev = self._next = self # sentinel node
        else:
            self._attach(prev) # non-sentinel node

        self.key = key
        self.value = value

    @property
    def prev(self):
        """The node before this one."""
        return self._prev

    @property
    def next(self):
        """The node after this one."""
        return self._next

    def connect_after(self, prev : '_Node'):
        """Places this node after prev. Updates prior neighbors."""
        self._prev._next = self._next
        self._next._prev = self._prev
        self._attach(prev)

    def _attach(self, prev : '_Node'):
        """Places this node after prev. Does not update prior neighbors."""
        self._prev = prev
        self._next = prev._next
        self._prev._next = self._next._prev = self


NOT_FOUND = -1


class LRUCache:
    """A cache with a least-recently-used validation policy."""

    __slots__ = ('_capacity', '_map', '_sentinel')

    def __init__(self, capacity: int):
        """Makes an initially empty LRU cache of the given capacity."""
        if capacity <= 0:
            raise ValueError('capacity must be positive')

        self._capacity = capacity
        self._map = {}
        self._sentinel = _Node()

    def get(self, key: int) -> int:
        """
        Retreives the value associated with the given key.
        If there is no mapping with that key, returns NOT_FOUND.
        """
        try:
            node = self._map[key]
        except KeyError:
            return NOT_FOUND

        self._bump(node)
        return node.value

    def put(self, key: int, value: int) -> None:
        """Associates a value with a (new or existing) key."""
        try:
            node = self._map[key]
        except KeyError:
            pass
        else:
            self._bump(node)
            node.value = value
            return

        if len(self._map) < self._capacity:
            node = _Node(key, value, self._sentinel)
        else:
            node = self._sentinel.prev
            del self._map[node.key]
            node.key = key
            node.value = value
            self._bump(node)

        self._map[key] = node

    def _bump(self, node):
        """Brings node to the front of the chain."""
        node.connect_after(self._sentinel)


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)
